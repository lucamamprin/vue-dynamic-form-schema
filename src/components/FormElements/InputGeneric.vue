<template>
  <div
    class="input floating-label"
    :class="[
      {
        'input-transparent': darkMode,
        'mt-0': resetMarginTop,
        'floating-label-floated': value || floatedLabel,
      },
      $attrs.type,
    ]"
  >
    <div class="position-relative">
      <input
        v-bind="$attrs"
        :class="{
          'border-danger': hasError,
        }"
        :aria-describedby="`${describedBy} ${maxLengthDescribedBy}`"
        :value="value"
        :name="name"
        :id="id"
        :disabled="disabled"
        :min="min"
        :max="max"
        :minlength="minLength"
        :maxlength="maxLength"
        :placeholder="placeholderText"
        :aria-labelledby="`label-${name}`"
        :aria-invalid="invalid.toString()"
        :aria-required="required.toString()"
        v-on="{
          ...$listeners,
          input: (event) =>
            $emit(
              'input',
              uppercase ? event.target.value.toUpperCase() : event.target.value
            ),
        }"
      >
      <label
        :id="`label-${name}`"
        :for="id"
        v-html="label"
      />
    </div>

    <div
      v-if="hasCharacterCount || hasLeftColSlot"
      class="pal-row pal-no-gutters mt-2"
    >
      <div
        v-if="hasLeftColSlot"
        class="pal-col px-0"
      >
        <slot name="leftCol" />
      </div>

      <div
        v-if="hasCharacterCount"
        class="pal-col px-0"
      >
        <max-length
          :typed-text="value"
          :sr-character-count="maxLengthDescribedBy"
          :label="removeStar(label)"
          :max-length="maxLength"
        />
      </div>
    </div>

    <div v-if="hasError && errorMessage">
      <error-message
        :id="describedBy"
        :error-message="errorMessage"
      />
    </div>
  </div>
</template>

<script>
import removeStar from "../../utilities/removeStar"
const MaxLength = () => import("./Atoms/MaxLength")
const ErrorMessage = () => import("./Atoms/ErrorMessage")

export default {
  name: "InputGeneric",
  inheritAttrs: false,
  mixins: [
    removeStar,
  ],
  components: {
    MaxLength,
    ErrorMessage,
  },
  computed: {
    hasLeftColSlot () {
      return this.$slots.leftCol
    },
    hasCharacterCount () {
      return this.maxLength && this.showCharacterCount
    },
    maxLengthDescribedBy () {
      return this.hasCharacterCount ? `${this.id}-counter` : ""
    },
  },
  props: {
    floatedLabel: {
      type: Boolean,
      default: false,
    },
    id: {
      type: String,
      required: true,
    },
    describedBy: {
      type: String,
      required: true,
    },
    value: {
      type: [
        String, Number,
      ],
      default: null,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    name: {
      type: String,
      required: true,
    },
    label: {
      type: String,
      required: true,
    },
    errorMessage: {
      type: String,
      default: null,
    },
    hasError: {
      type: Boolean,
      default: false,
    },
    min: {
      type: null,
      default: null,
    },
    max: {
      type: null,
      default: null,
    },
    minLength: {
      type: Number,
      default: null,
    },
    maxLength: {
      type: Number,
      default: null,
    },
    showCharacterCount: {
      type: Boolean,
      default: true,
    },
    resetMarginTop: {
      type: Boolean,
      default: false,
    },
    darkMode: {
      type: Boolean,
      default: false,
    },
    placeholderText: {
      type: String,
      default: " ",
    },
    invalid: {
      type: Boolean,
    },
    required: {
      type: Boolean,
      default: false,
    },
    uppercase: {
      type: Boolean,
      default: false,
    },
  },
}
</script>
