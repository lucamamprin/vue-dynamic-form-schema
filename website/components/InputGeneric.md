# InputGeneric

## Props

<!-- @vuese:InputGeneric:props:start -->
|Name|Description|Type|Required|Default|
|---|---|---|---|---|
|value|-|`String` /  `Number`|`false`|null|
|type|-|`String`|`false`|text|
|min|-|—|`false`|null|
|max|-|—|`false`|null|

<!-- @vuese:InputGeneric:props:end -->


## Slots

<!-- @vuese:InputGeneric:slots:start -->
|Name|Description|Default Slot Content|
|---|---|---|
|leftCol|-|-|

<!-- @vuese:InputGeneric:slots:end -->


## MixIns

<!-- @vuese:InputGeneric:mixIns:start -->
|MixIn|
|---|
|props|
|textInputs|

<!-- @vuese:InputGeneric:mixIns:end -->


