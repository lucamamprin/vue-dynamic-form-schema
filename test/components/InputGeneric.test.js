import Vue from "vue"
import { mount } from "@vue/test-utils"
import InputGeneric from "../../src/components/FormElements/InputGeneric.vue"

const props = {
  id: "firstname",
  describedBy: "firstname-describe",
  name: "firstname",
  label: "First name",
  required: true,
}

let wrapper 

describe("InputGeneric", () => {
  beforeEach(() => {
    wrapper = mount(InputGeneric, {
      propsData: props,
    })
  })

  test("mounts correctly", () => {
    expect(wrapper.vm).toBeTruthy()
    expect(wrapper.html()).toMatchSnapshot()
  })

  test("displays proper value if already set", async () => {
    await wrapper.setProps({
      value: "John",
    })

    expect(wrapper.find("input").element.value).toBe("John")
  })

  test("prints the floating label classes if user types a value", async () => {
    await wrapper.setProps({
      value: "John",
    })

    expect(wrapper).toMatchSnapshot()
  })

  test("prints error message", async () => {
    await wrapper.setProps({
      hasError: true,
      errorMessage: "Generic error message",
    })

    await Vue.nextTick()

    expect(wrapper).toMatchSnapshot()
  })
})
