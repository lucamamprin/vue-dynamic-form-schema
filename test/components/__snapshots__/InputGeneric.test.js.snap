// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input Generic - input number has max and min values 1`] = `
<div class="input floating-label">
  <div class="position-relative"><input type="text" aria-describedby="age-describe " name="age" id="age-input" placeholder=" " aria-labelledby="label-age" aria-invalid="false" aria-required="false" class="" min="20" max="50"> <label id="label-age" for="age-input">Age</label></div>
  <!---->
  <!---->
</div>
`;

exports[`Input Generic - input number number mounts correctly 1`] = `
<div class="input floating-label">
  <div class="position-relative"><input type="text" aria-describedby="age-describe " name="age" id="age-input" placeholder=" " aria-labelledby="label-age" aria-invalid="false" aria-required="false" class=""> <label id="label-age" for="age-input">Age</label></div>
  <!---->
  <!---->
</div>
`;

exports[`InputGeneric mounts correctly 1`] = `
<div class="input floating-label">
  <div class="position-relative"><input type="text" aria-describedby="firstname-describe " name="firstname" id="firstname" placeholder=" " aria-labelledby="label-firstname" aria-invalid="false" aria-required="true" class=""> <label id="label-firstname" for="firstname">First name</label></div>
  <!---->
  <!---->
</div>
`;

exports[`InputGeneric prints error message 1`] = `
<div class="input floating-label">
  <div class="position-relative"><input type="text" aria-describedby="firstname-describe " name="firstname" id="firstname" placeholder=" " aria-labelledby="label-firstname" aria-invalid="false" aria-required="true" class="border-danger"> <label id="label-firstname" for="firstname">First name</label></div>
  <!---->
  <div><span id="firstname-describe" class="help m-0   d-inline-block">
  Generic error message
</span></div>
</div>
`;

exports[`InputGeneric prints the floating label classes if user types a value 1`] = `
<div class="input floating-label floating-label-floated">
  <div class="position-relative"><input type="text" aria-describedby="firstname-describe " name="firstname" id="firstname" placeholder=" " aria-labelledby="label-firstname" aria-invalid="false" aria-required="true" class=""> <label id="label-firstname" for="firstname">First name</label></div>
  <!---->
  <!---->
</div>
`;

exports[`InputGeneric show minlength and maxlength attributes and component 1`] = `
<div class="input floating-label">
  <div class="position-relative"><input type="text" aria-describedby="firstname-describe firstname-counter" name="firstname" id="firstname" placeholder=" " aria-labelledby="label-firstname" aria-invalid="false" aria-required="true" class="" minlength="20" maxlength="50"> <label id="label-firstname" for="firstname">First name</label></div>
  <div class="pal-row pal-no-gutters mt-2">
    <!---->
    <div class="pal-col px-0">
      <div class="text-right text-light"><span id="firstname-counter" class="sr-only">
    0 characters on 50 available for First name
  </span>
        <p class="m-0">
          0 / 50
        </p>
      </div>
    </div>
  </div>
  <!---->
</div>
`;

exports[`multiple input types email 1`] = `
<div class="input floating-label">
  <div class="position-relative"><input type="email" aria-describedby="email-describe " name="email-name" id="email-input" placeholder=" " aria-labelledby="label-email-name" aria-invalid="false" aria-required="false" class=""> <label id="label-email-name" for="email-input">email-label</label></div>
  <!---->
  <!---->
</div>
`;

exports[`multiple input types password 1`] = `
<div class="input floating-label">
  <div class="position-relative"><input type="password" aria-describedby="password-describe " name="password-name" id="password-input" placeholder=" " aria-labelledby="label-password-name" aria-invalid="false" aria-required="false" class=""> <label id="label-password-name" for="password-input">password-label</label></div>
  <!---->
  <!---->
</div>
`;

exports[`multiple input types search 1`] = `
<div class="input floating-label">
  <div class="position-relative"><input type="search" aria-describedby="search-describe " name="search-name" id="search-input" placeholder=" " aria-labelledby="label-search-name" aria-invalid="false" aria-required="false" class=""> <label id="label-search-name" for="search-input">search-label</label></div>
  <!---->
  <!---->
</div>
`;

exports[`multiple input types tel 1`] = `
<div class="input floating-label">
  <div class="position-relative"><input type="tel" aria-describedby="tel-describe " name="tel-name" id="tel-input" placeholder=" " aria-labelledby="label-tel-name" aria-invalid="false" aria-required="false" class=""> <label id="label-tel-name" for="tel-input">tel-label</label></div>
  <!---->
  <!---->
</div>
`;
